swarm-test/src/lib.rs

```diff
@@ -0,0 +1,353 @@
+/// For example, if you expect the first [`Swarm`] to emit 2 events, you should assign the first variable of the returned tuple value to an array of size 2.
+/// This works especially well if you directly pattern-match on the return value.
+///
+/// ## Type of event
+///
+/// This function utilizes the [`TryIntoOutput`] trait.
+/// Similar as to the number of expected events, the type of event is inferred based on your usage.
+/// If you match against a [`SwarmEvent`], the first [`SwarmEvent`] will be returned.
+/// If you match against your [`NetworkBehaviour::OutEvent`] type, [`SwarmEvent`]s which are not [`SwarmEvent::Behaviour`] will be skipped until the [`Swarm`] returns a behaviour event.
+///
+/// You can implement the [`TryIntoOutput`] for any other type to further customize this behaviour.
+///
+/// # Difference to [`futures::future::join`]
+///
+/// This function is similar to joining two futures with two crucial differences:
+/// 1. As described above, it allows you to obtain more than a single event.
+/// 2. More importantly, it will continue to poll the [`Swarm`]s **even if they already has emitted all expected events**.
+///
+/// Especially (2) is crucial for our usage of this function.
+/// If a [`Swarm`] is not polled, nothing within it makes progress.
+/// This can "starve" the other swarm which for example may wait for another message to be sent on a connection.
+///
+/// Using [`drive`] instead of [`futures::future::join`] ensures that a [`Swarm`] continues to be polled, even after it emitted its events.
+pub async fn drive<
+    TBehaviour1,
+    const NUM_EVENTS_SWARM_1: usize,
+    Out1,
+    TBehaviour2,
+    const NUM_EVENTS_SWARM_2: usize,
+    Out2,
+>(
+    swarm1: &mut Swarm<TBehaviour2>,
+    swarm2: &mut Swarm<TBehaviour1>,
+) -> ([Out1; NUM_EVENTS_SWARM_1], [Out2; NUM_EVENTS_SWARM_2])
+where
+    TBehaviour2: NetworkBehaviour + Send,
+    TBehaviour2::OutEvent: Debug,
+    TBehaviour1: NetworkBehaviour + Send,
+    TBehaviour1::OutEvent: Debug,
+    SwarmEvent<TBehaviour2::OutEvent, THandlerErr<TBehaviour2>>: TryIntoOutput<Out1>,
+    SwarmEvent<TBehaviour1::OutEvent, THandlerErr<TBehaviour1>>: TryIntoOutput<Out2>,
+    Out1: Debug,
+    Out2: Debug,
+{
+    let mut res1 = Vec::<Out1>::with_capacity(NUM_EVENTS_SWARM_1);
+    let mut res2 = Vec::<Out2>::with_capacity(NUM_EVENTS_SWARM_2);
+
+    while res1.len() < NUM_EVENTS_SWARM_1 || res2.len() < NUM_EVENTS_SWARM_2 {
+        match futures::future::select(swarm1.next_swarm_event(), swarm2.next_swarm_event()).await {
+            Either::Left((o1, _)) => {
+                if let Ok(o1) = o1.try_into_output() {
+                    res1.push(o1);
+                }
+            }
+            Either::Right((o2, _)) => {
+                if let Ok(o2) = o2.try_into_output() {
+                    res2.push(o2);
+                }
+            }
+        }
+    }
```

Comment:
Long term it would be nice to detect missed wake-ups in a `NetworkBehaviour` implementation with `libp2p-swarm-test`. If I am not mistaken, that is not possible with the above `while` `select!` loop as long as the other swarm triggers a wake-up.

Lets defer to a future pull request.

Reply:
This is kind of a conflicting requirement. I added this because it is very easy to poll a `Swarm` in a wrong way in a test, thus it is often the _test_ that is buggy, not the `NetworkBehaviour` implementation.